diff --git a/Makefile b/Makefile
index 1a56721..76d6ead 100644
--- a/Makefile
+++ b/Makefile
@@ -4,7 +4,7 @@
 BINDINGS_DIR = bindings
 CORE_MODULES = glib gobject gio gmodule atk gdkpixbuf freetype harfbuzz cairo pango xlib types/libc
 MODULES3 = $(CORE_MODULES) gdk gtk
-MODULES4 = $(CORE_MODULES) gdk4 gtk4 pangocairo graphene gsk
+MODULES4 = $(CORE_MODULES) gdk4 gtk4 pangocairo graphene gsk adw
 
 DESTDIR =
 PREFIX = /usr/local
diff --git a/format/fastxml/parser.ha b/format/fastxml/parser.ha
index f970ac5..5dc1b52 100644
--- a/format/fastxml/parser.ha
+++ b/format/fastxml/parser.ha
@@ -369,7 +369,7 @@ fn scan_charref(par: *parser) (rune | error) = {
 	if (len(memio::string(&par.entbuf)!) == 0) {
 		return par.line: syntaxerr;
 	};
-	match (strconv::stou32b(memio::string(&par.entbuf)!, base)) {
+	match (strconv::stou32(memio::string(&par.entbuf)!, base)) {
 	case let u: u32 =>
 		return u: rune;
 	case (strconv::invalid | strconv::overflow) =>
diff --git a/generate4 b/generate4
index de2eb6f..85b26d1 100755
--- a/generate4
+++ b/generate4
@@ -5,7 +5,7 @@ BIND=./bindings
 mkdir -p "$BIND"/glib "$BIND"/gobject "$BIND"/gio "$BIND"/gmodule "$BIND"/atk \
 	"$BIND"/gdkpixbuf "$BIND"/freetype "$BIND"/harfbuzz "$BIND"/cairo \
 	"$BIND"/pango "$BIND"/pangocairo "$BIND"/gdk4 "$BIND"/xlib "$BIND"/graphene \
-	"$BIND"/gsk "$BIND"/gtk4
+	"$BIND"/gsk "$BIND"/gtk4 "$BIND"/adw
 
 hare build cmd/hare-gi && exec ./hare-gi -B \
 	-m GLib,glib,"$BIND"/glib/_generated.ha \
@@ -24,6 +24,7 @@ hare build cmd/hare-gi && exec ./hare-gi -B \
 	-m Graphene,graphene,"$BIND"/graphene/_generated.ha \
 	-m Gsk,gsk,"$BIND"/gsk/_generated.ha \
 	-m Gtk,gtk4,"$BIND"/gtk4/_generated.ha \
+	-m Adw,adw,"$BIND"/adw/_generated.ha \
 	/usr/share/gir-1.0/GLib-2.0.gir \
 	/usr/share/gir-1.0/GObject-2.0.gir \
 	/usr/share/gir-1.0/Gio-2.0.gir \
@@ -39,4 +40,5 @@ hare build cmd/hare-gi && exec ./hare-gi -B \
 	/usr/share/gir-1.0/xlib-2.0.gir \
 	/usr/share/gir-1.0/Graphene-1.0.gir \
 	/usr/share/gir-1.0/Gsk-4.0.gir \
-	/usr/share/gir-1.0/Gtk-4.0.gir
+	/usr/share/gir-1.0/Gtk-4.0.gir \
+	/usr/share/gir-1.0/Adw-1.gir
diff --git a/gir/parse.ha b/gir/parse.ha
index 91c55d4..84198de 100644
--- a/gir/parse.ha
+++ b/gir/parse.ha
@@ -439,7 +439,7 @@ fn parse_member(parser: *xml::parser, members: *opaque) (void | error) = {
 };
 
 // callable
-fn parse_callable(callable: *callable) [17]scan = [
+fn parse_callable(callable: *callable) [20]scan = [
 	("introspectable", null),
 	("deprecated", null),
 	("deprecated-version", null),
@@ -457,6 +457,9 @@ fn parse_callable(callable: *callable) [17]scan = [
 	("shadows", &callable.shadows),
 	("throws", &callable.throws),
 	("moved-to", null),
+	("glib:async-func", null),
+	("glib:sync-func", null),
+	("glib:finish-func", null)
 ];
 fn parse_parameters(parser: *xml::parser, parameters: *opaque) (void | error) = {
 	return parse_element(parser, "parameters",
